src/tap/grpc/server.rs

```diff
@@ -255,24 +284,70 @@ impl iface::Tap for Tap {
                     base: shared.base_id,
                     stream: next_id as u64,
                 };
-                Some((id, shared.events_tx.clone()))
-            } else {
-                None
+                // Is HTTP data being extracted from the request, and should
+                // headers be included?
+                match shared.extract {
+                    Extract::Http { headers } => return Some((id, shared.events_tx.clone(), headers)),
```

Comment:
So, this means that `Extract::Http` _must_ always be sent now, correct? While I agree with the intent, this is backwards-incompatible, so we should avoid it.  I think we should basically make this `let extract = shared.extract.map(|e| match e { Extract::Http { headers} => headers, _ => false }).unwrap_or(false)` ish thing... i.e. so that we only use this field to determine whether or not headers should be extracted. We can revisit the defaults later once we've reasonably expected all clients to take the new API.

Reply:
@olix0r That's not actually the case; `shared.extract` is a local type, _not_ the deserialized protobuf type: 
https://github.com/linkerd/linkerd2-proxy/blob/4b13908bad52f3d2aca84bebf4b941576f27fd29/src/tap/grpc/server.rs#L88-L91
which we construct from the received protobuf to make the representation simpler to match against internally.

If we receive an observe request with no `extract` field, we'll just construct `Extract::Http { headers: false }` by default, to preserve backwards compatibility:
https://github.com/linkerd/linkerd2-proxy/blob/4b13908bad52f3d2aca84bebf4b941576f27fd29/src/tap/grpc/server.rs#L140-L152

Reply:
OK, I think I understand. It's a little surprising given the overlapping names but I get where you're going. I think `Extract` could use a docstring explaining why it exists and how it relates to the protobuf concept of Extract. A name change could also help clarify (e.g., `ShouldExtractHttp`? idk) perhaps

Reply:
Okay, I can add some docs — I was basing this off of the similar pattern in the match module, where we have a separate internal match type that's constructed from the protobuf match type.